File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        ListNode curr1 = list1;
        ListNode curr2 = list2;

        if (list1 == null && list2 == null) {
            return null;
        }
        else if (list1 == null) {
            return list2;
        }
        else if (list2 == null) {
            return list1;
        }


        ListNode currRes = null;
        if (curr1.val < curr2.val) {
            currRes = curr1;
            curr1 = curr1.next;
        } else {
            currRes = curr2;
            curr2 = curr2.next;
        }
        ListNode resHead = currRes;

        while (!(curr1 == null || curr2 == null)) {
            if (curr1.val < curr2.val) {
                currRes.next = curr1;
                curr1 = curr1.next;
                currRes = currRes.next;
            } else {
                currRes.next = curr2;
                curr2 = curr2.next;
                currRes = currRes.next;

            }
        }

        while (curr1 != null) {
            currRes.next = curr1;
            curr1 = curr1.next;
            currRes = currRes.next;
        }
        while (curr2 != null) {
            currRes.next = curr2;
            curr2 = curr2.next;
            currRes = currRes.next;
        }

        return resHead;


    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:

        # find midpoint: slow will land on last element in first half
        fast = head
        slow = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        # store ref so we can iterate after we reverse second half
        back = slow.next
        slow.next = None  # separate

        # reverse ts
        prev = None
        while back:
            next = back.next
            back.next = prev
            prev = back
            back = next
        back = prev

        # combine these jawns
        
        front = head
        while front and back:  # front will reach null, as slow.next is now null
            nextFront, nextBack = front.next, back.next

            front.next = back
            back.next = nextFront

            front = nextFront
            back = nextBack

        # front will always exhaust, back may have a leftover
        if back:
            front.next = back

        
